feat(mobile): show more detail in list view

safwansamsudeen opened the pull request:
> Please provide enough information so that others can review your pull request:

Closes #14685, and is part of #18921.

> Explain the **details** for making this change. What existing problem does the pull request solve?

Previously, only two fields were visible on mobile. This PR will give a two-layered row - with the Subject at the top, and up to 3 fields below.

There are many types of columns, and this PR might not work everywhere. It definitely will break anything, but it might result in weird output on mobile. We'll fix these as we encounter them.

> Screenshots/GIFs

Here's how it looks now:
<img width="483" alt="image" src="https://github.com/frappe/frappe/assets/62411302/50f0a2a6-1425-4fe5-900e-da974b293512">

<img width="497" alt="image" src="https://github.com/frappe/frappe/assets/62411302/3f77e70b-b54f-460e-8724-2b135b6f0662">



`no-docs`

safwansamsudeen committed:
feat(mobile): show more detail in list view

Olawale1 commented:
Hi @safwansamsudeen 

Thanks for the great job so far. How does the user select which fields are displayed ?

Kind regards,

safwansamsudeen commented:
@Olawale1 this isn't implemented right now. Perhaps in a follow up PR?

Olawale1 commented:
> @Olawale1 this isn't implemented right now. Perhaps in a follow up PR?

Okay, that is quite important because different users have different priorities 

But how does it work now ? Does it just pick the first set of fields with 'in list view' checked ?

safwansamsudeen commented:
@Olawale1 Yeah, it picks up the first three fields that have a type of `Field`. If there isn't three, it picks up as much as possible.

ankush commented:
User wise configurable columns are a separate feature IMO. Even desktop view doesn't support it, so lets first focus on this :+1:

Olawale1 commented:
> @Olawale1 Yeah, it picks up the first three fields that have a type of `Field`. If there isn't three, it picks up as much as possible.

I'm not sure what you mean by 'a type of field' but if it's a fieldtype that's dedicated for this purpose then the requirement is already met ! If it's the 'in list view' field then that's manageable for now though required columns for Desk and Mobile could often differ

Many thanks

Olawale1 commented:
> User wise configurable columns are a separate feature IMO. Even desktop view doesn't support it, so lets first focus on this 👍

Hi Ankush,

Of course desktop view supports this... that's the purpose of the 'in list view' option... that's how the user (System Manager / Administrator) decides which columns should be visible in List / Grid view ! It is important to also have such an option for Mobile List View

Thanks plenty

ankush commented:
@Olawale1 that same thing works in mobile too, it's not mobile specific configuration. 

@safwansamsudeen can you confirm first first 3 fields are picked based on same configuration? :thinking:

maharshivpatel commented on frappe/public/js/frappe/list/list_view.js:
```diff
@@ -697,7 +697,18 @@ frappe.views.ListView = class ListView extends frappe.views.BaseList {
 
 	get_left_html(doc) {
-		return this.columns.map((col) => this.get_column_html(col, doc)).join("");
+		let field_columns = this.columns.filter(({ type }) => type === "Field").slice(0, 4);
```
@safwansamsudeen  As we are already adding columns in get_column_html, they are just [hidden](https://github.com/frappe/frappe/blob/e4dd402e294e5a34e9827b6524a12b90aa26c756/frappe/public/js/frappe/list/list_view.js#L838-L841) on mobile. can we modify get_column_html function it self ?

safwansamsudeen replied:
@maharshivpatel good idea, I'll do that.

safwansamsudeen replied:
@maharshivpatel what I'm doing, actually, is to give an alternative view with `mobile_row`... and the first one is hidden.

`get_column_html` gives the html for that one _column_, while I attempt to give an alternative for the whole left html.

maharshivpatel replied:
@safwansamsudeen you also have to format the values
## Raw
<img width="431" alt="Screenshot 2024-01-29 at 4 51 56 PM" src="https://github.com/frappe/frappe/assets/39730881/0db135a3-df26-4b83-b20d-a57754dcd81d">



## Formatted
<img width="431" alt="Screenshot 2024-01-29 at 4 47 59 PM" src="https://github.com/frappe/frappe/assets/39730881/8c49376b-9c13-43e9-a78c-4a74ee810de4">

safwansamsudeen commented:
@ankush @Olawale1 yes, it does. Modified Activity Log view:
<img width="470" alt="image" src="https://github.com/frappe/frappe/assets/62411302/fe47dfed-f450-4e47-9044-62dda9e73fc4">


(Sorry for the delay, this slipped through the cracks).

maharshivpatel commented on frappe/public/scss/desk/list.scss:
```diff
@@ -81,6 +81,29 @@
 	@include get_textstyle("base", "regular");
 	border-radius: var(--border-radius-md);
 
+	@media (max-width: 768px) {
```
```suggestion
	@media (max-width: map-get($grid-breakpoints, "md")) {
```

safwansamsudeen replied:
This doesn't work for me some reason. Does it work for you?

maharshivpatel commented on frappe/public/scss/desk/list.scss:
```diff
@@ -81,6 +81,29 @@
+		height: fit-content;
+		padding-top: 4.5px;
+		padding-bottom: 4.5px;
+
```
let's decrease font size as well ?
```suggestion
		@include get_textstyle("2xs", "regular");

```

safwansamsudeen replied:
It feels a little too small to me:
<img width="377" alt="image" src="https://github.com/frappe/frappe/assets/62411302/c4e42579-a463-48c9-9724-a598852c9d80">

What about `md`? It's 13px, while this is 12. Here, the subject is `md` while the bottom is `sm`.
<img width="397" alt="image" src="https://github.com/frappe/frappe/assets/62411302/3066c99a-c5a1-43fc-b344-7e0c6dbe351a">

maharshivpatel committed:
fix: mobile view modify instead of adding columns

As we are already have columns for desktop view.
we just need to modify them for mobile view instead of adding new columns.

maharshivpatel committed:
Merge branch 'develop' into mobile-list-row

maharshivpatel committed:
Merge branch 'develop' into mobile-list-row

safwansamsudeen reviewed (commented):
Sorry for the delay, was working on another project. Thanks for your help @maharshivpatel!

safwansamsudeen committed:
fix: apply suggestions from code review

Co-authored-by: Maharshi Patel <39730881+maharshivpatel@users.noreply.github.com>

safwansamsudeen committed:
Merge branch 'mobile-list-row' into fix-reuse-columns-for-mobile

safwansamsudeen committed:
merge: mobile view modify instead of adding columns

ankush commented:
@safwansamsudeen the test failures are likely related. See list view ones, I'll check the others. 

Test file is ins cypress/integration if you can reproduce locally that most likely means something broke.

maharshivpatel commented:
checkbox was moved out main row div container that must have broken few test/functionality

ankush committed:
Merge branch 'develop' into mobile-list-row

ankush committed:
Merge branch 'develop' into mobile-list-row

ankush committed:
test: fix selector for checkbox

ankush committed:
fix: Ignore stderr

ankush committed:
fix: selector for drag select

ankush commented:
@safwansamsudeen last bit of alignment and UI problems left

![image](https://github.com/frappe/frappe/assets/9079960/0eae27f7-2183-445d-8c9b-c9ebdf141f44)

safwansamsudeen committed:
fix: indicator disappearing

safwansamsudeen committed:
fix: alignment

safwansamsudeen commented:
Alignment is fixed (before/after):
<img width="1418" alt="image" src="https://github.com/frappe/frappe/assets/62411302/d198702e-6c91-408d-8dfa-fc87037888c8">

Disappearing indicator was fixed.

safwansamsudeen committed:
fix: tests

safwansamsudeen committed:
fix: tests again

safwansamsudeen committed:
fix: make button actually work

ankush commented:
@mergifyio update

ankush committed:
Merge branch 'develop' into mobile-list-row

ankush commented:
@safwansamsudeen the alignment is still kinda broken :smiling_face_with_tear: 
before:
![image](https://github.com/frappe/frappe/assets/9079960/a74af94b-8c8f-483b-9888-3280637e6c14)


after:
![image](https://github.com/frappe/frappe/assets/9079960/55ff7ac1-1be7-4945-9784-29309c7c2eae)

safwansamsudeen committed:
fix: column alignment

safwansamsudeen commented:
@ankush any reason not to merge this?

ankush commented:
Closing since multiple UI inconsistencies still and there's parallel project going on for UI rewrite

ankush commented:
@iamejaaz if you want to attempt this again. I'm sure people will love it. List view revamp is one of the most request feature (not just for mobile view)
